chain-core/src/common/merkle_tree.rs

```diff
@@ -808,4 +809,17 @@ mod tests {
             .unwrap()
             .verify(&new_tree.root_hash()));
     }
+
+    #[test]
+    fn check_wrong_leaf_value() {
+        let values = vec!["one", "two", "three", "four"];
+        let tree = MerkleTree::new(values);
+
+        let mut proof = tree.generate_proof("one").unwrap();
+        assert!(proof.verify(&tree.root_hash()));
+
+        // Intentionally change the value in proof
+        proof.value = "two";
```

Comment:
Note that it is not possible write this line of code from outside of this module because value is a private member of `Proof`. So, there is no way to change the value of a proof or create an incorrect proof in current public API.

Reply:
it's possible with "encode" and directly manipulating the bytes

Reply:
Yeah. Anyway, we have this check in place now.